scipy/sparse/linalg/isolve/minres.py

```diff
@@ -110,23 +110,19 @@ def minres(A, b, x0=None, shift=0.0, tol=1e-5, maxiter=None,
 
     eps = finfo(xtype).eps
 
-    x = zeros(n, dtype=xtype)
-
     # Set up y and v for the first Lanczos vector v1.
     # y  =  beta1 P' v1,  where  P = C**(-1).
     # v is really P' v1.
 
-    y = b
-    r1 = b
-
-    y = psolve(b)
+    r1 = b - A*x
+    y = psolve(r1)
 
-    beta1 = inner(b,y)
+    beta1 = inner(y, y)
```

Comment:
Did you check that this change is correct? It's now different from previous, also for `x0=0`.

Reply:
I was indeed to focused on no preconditioning, let me get back to it and start by adding more tests!

Reply:
Corrected to a version that works also with preconditioning and added a test

Reply:
And then a few more questions: do you understand the minres algorithm? Did you check that these changes are correct (without assuming the previous PRs are)?

Reply:
I have quite some experience and understanding of (lanczos bases) optimisation algorithms. However, in practice, all is limited to non-preconditioned systems!

In a Lanczos based approach, if you have an initial vector, basically all you need to change is the initial residual, which happens on line 117 in the PR. This is also compatible with the way the original code deals with next residuals: https://github.com/scipy/scipy/blob/4d26a150340a51153b01615c8afbe433380340d8/scipy/sparse/linalg/isolve/minres.py#L196
Notice that the code now also reproduces exactly what happened before in case `x0=0`.
I looked in the literature and as far as I can tell, this introduction of `x0` is indeed correct. However, as I said, I have no extended practical experience with preconditioners and given that I'm working on someone else code, I cannot say I'm 100% sure that there is subtle mistake, sorry.

Reply:
Thanks, was just asking before thinking it through myself. After reading it properly, the changes look ok.